### ISSUE TITLE
Tensor print format issue on Nvidia Jetson devices

### ISSUE THREAD

**Issue by yindavidyang** [pytorch/pytorch#49452]
## To Reproduce

1. Use a Jetson device such as AGX Xavier, and install the latest OS (from Jetpack 4.4.1)
2. Download and install the official Pytorch 1.7 build from Nvidia: https://nvidia.box.com/shared/static/wa34qwrwtk9njtyarwt5nvo6imenfy26.whl
3. run the following:

import torch
x = torch.tensor([1.234])
print(x)

The result is 1., while it should be 1.234. 

The problem is for CPU tensor only. print(x.cuda()) gives the correct result, but print(x.cuda().cpu()) gives the above problmatic result. Also, print(x.numpy()) gives the correct result.

v1.7.1 has the same issue, but earlier versions such as v1.6 is does not.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.7.0 and 1.7.1
 - OS (e.g., Linux): Nvidia L4T 4.4.1
 - How you installed PyTorch (`conda`, `pip`, source): source, nvidia official build
 - Build command you used (if compiling from source): following https://forums.developer.nvidia.com/t/pytorch-for-jetson-version-1-7-0-now-available/72048
 - Python version: 3.6
 - CUDA/cuDNN version: 10.2/8.0
 - GPU models and configuration: AGX Xavier
 - Any other relevant information:

## Additional context

cc @ezyang @gchanan @zou3519 @bdhirsh @seemethere @malfet @walterddr @ngimel

**Comment by ezyang** [comment]
Is this a printing issue or an actual numeric issue? (e.g., if you extract out the value with x.item() what does it print as?)

**Comment by yindavidyang** [comment]
It's a numeric issue, not just a printing issue. Both x.item() and x.numpy() give the correct results. However:

import torch
import torch.nn.functional as F
x = torch.tensor(12.345)
print(x)
print(F.softmax(x))

Expected results:
12.345
1.

Actual output:
12.
nan

Note that if I do `x = x.cuda()` before the printing, then the results are correct. Also, this issue doesn't happen to v1.6.0, or to the Desktop version of PyTorch 1.7.

**Comment by ezyang** [comment]
This is pretty wild. Since the NVIDIA binaries are involved, I'm going to tag CUDA (although this is a CPU only repro). Also tagging binaries.

**Comment by malfet** [comment]
I suspect it's the same NEON issue, namely https://github.com/pytorch/pytorch/pull/49273
Can not reproduce this issue using torch_nighly builds:
```
$ pip install --pre torch  -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
$ python -c "import torch;x=torch.tensor(12.345);print(x, torch.nn.functional.softmax(x))"
-c:1: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
tensor(12.3450) tensor(1.)
```

**Comment by yindavidyang** [comment]
Tried #49273 and it doesn't work. Torch nightly doesn't have this issue, but it's limited to CPU with no CUDA support.

Here's why (I guess):

- #49273 fixes missing_vst1_neon.h. This file is only compiled when MISSING_ARM_VST1 is defined, and MISSING_ARM_VLD1 is not defined.
- I guess Torch nightly is built with GCC8, which leads to MISSING_ARM_VST1, but not MISSING_ARM_VLD1. So it doesn't encounter this issue thanks to #49273.
- Nvidia's OS uses GCC7, which leads to both MISSING_ARM_VST1 and MISSING_ARM_VLD1. (Not sure we can change that to GCC8 since CUDA10.2 was built with GCC7.)
- MISSING_ARM_VLD1 is addressed in missing_vld1_neon.h, which has the old unfixed version of vst1q_f32_x2.

So here's what I suggest:

- Modify every vstq1_* function in missing_vld1_neon.h similarly as #49273.

**Comment by yindavidyang** [comment]
Tried the master branch code on Jetson (GCC7), and the results are correct. 

Compared to v1.7.1, a major change in the master branch is that vec256_float_neon.h has a new compiler check and only compiles for clang and GCC>8.3. So, for GCC7, the code that causes this issue is skipped.

So this issue is kind of already fixed in the master branch ... by essentially disabling neon for Jetson.

**Comment by malfet** [comment]
@yindavidyang I have a longer patch from @sebpop that should fix compilation problems with gcc-7, will try to merge it soon.

**Comment by yindavidyang** [comment]
I informed nvidia, and the moderator fixed the pytorch wheel. thanks @malfet for the help.

**Comment by yindavidyang** [comment]
> @yindavidyang I have a longer patch from @sebpop that should fix compilation problems with gcc-7, will try to merge it soon.

Look forward to the new patch -- without NEON, CPU operations are very slow on Jetson.
